### PR TITLE
fix(v1/spec-decode): apply min_p to verified target tokens

### LINKED ISSUE #42744: [Bug]: rejection sampler skips argmax-invariant procs — patching the min_p/logit_bias 400 alone is insufficient
The guard at `vllm/sampling_params.py:_validate_spec_decode` 400s any spec-decode request with `min_p > _SAMPLING_EPS` or `logit_bias` set (added in #31982). Patching the raise out locally to let the params through reveals the underlying gap: the rejection sampler at `vllm/v1/sample/rejection_sampler.py:apply_logits_processors` only iterates `sampling_metadata.logitsprocs.non_argmax_invariant`. Both `MinPLogitsProcessor.is_argmax_invariant()` and `LogitBiasLogitsProcessor.is_argmax_invariant()` return `True` — the comment in `vllm/v1/sample/logits_processor/builtin.py` even says "Min-p never impacts greedy sampling".

Net: even with the guard removed, min_p / logit_bias are NOT applied to verified target-logit tokens during rejection. They'd only land on bonus tokens via the regular Sampler path. Inconsistent honoring across a single response, arguably worse than the current 400 if a caller assumed they'd taken effect.

Two ways to close this properly:

1. Mirror the regular Sampler in `apply_logits_processors` and include argmax-invariant procs (preserves determinism for greedy callers via the existing sampling_metadata check).
2. Route requests with `min_p`/`logit_bias` set to a non-spec target when they're present, leaving the guard as a hard fallback.

## Repro

```
vllm serve <model> --speculative-config '{"method":"ngram","num_speculative_tokens":5}'
curl /v1/chat/completions -d '{"messages":[...], "min_p":0.05}'
# 400 BadRequest: "The min_p and logit_bias sampling parameters are not yet supported with speculative decoding."
```

## Context

Hitting this running `RedHatAI/Qwen3-Coder-Next-NVFP4` with ngram speculative decoding on a DGX Spark (GB10 / sm_121a) via the Avarok-Cybersecurity/dgx-vllm fork. Qwen-family AWQ/NVFP4 leans on `min_p` for repetition mitigation on code tasks -- without it (or equivalent), our diff-format outputs malform at ~80% rate. `presence_penalty` / `frequency_penalty` are accepted via #19482 and partially compensate, but they aren't equivalent for this failure mode.

## Related

- #31982 (the guard PR — added the 400)
- #19482 (logit procs wired into rejection sampler — bad_words / penalties / allowed_token_ids only)
- #18902 (older tracking issue, closed stale)

### PR DESCRIPTION
**Pull request by camerono**
### what

`RejectionSampler.apply_logits_processors` only iterates `sampling_metadata.logitsprocs.non_argmax_invariant`, which excludes `MinPLogitsProcessor` (min_p is argmax-invariant). The consequence is that under speculative decoding, min_p is silently dropped on every verified target token — only the bonus token via the regular `Sampler` path honors it.

`SamplingParams._validate_spec_decode` was added to 400 these requests precisely because the behavior was partial. But that guard is the symptom; the actual fix is to apply min_p to the target tokens. This PR does that.

### the change

1. New method `MinPLogitsProcessor.apply_with_spec_decode(logits, num_draft_tokens)` — same softmax / max-prob / threshold / mask body as `apply()`, but expands the per-request `self.min_p` tensor to per-row using `repeat_interleave`, matching the layout of the spec-decode logits tensor (`[sum(num_draft_tokens), V]`).
2. `RejectionSampler.apply_logits_processors` gets a second iteration loop over `logitsprocs.argmax_invariant`, dispatching to `MinPLogitsProcessor.apply_with_spec_decode` (mirrors the existing `MinTokensLogitsProcessor` dispatch).
3. New unit tests in `tests/v1/sample/test_min_p_spec_decode.py`: no-op when disabled, per-request masking via `repeat_interleave`, mixed zero / non-zero `min_p` peers.

`_validate_spec_decode` is left in place — the validator should be relaxed in a follow-up once `logit_bias` is also wired (separate; not in this PR).

### why it matters

Without min_p under spec decoding, Qwen3-Coder-Next 80B-A3B (and similar code-tuned models) hit repetition collapse on long completions — the model emits `      result      result      result …` until the cap. With min_p applied to target logits via this patch, that disappears.

Measured on Qwen3-Coder-Next NVFP4 (vLLM 0.20.0 with this patch + Avarok kernel) on a single GB10:

| | spec off, min_p applied | spec on, no min_p | spec on, **this patch** |
|---|---|---|---|
| 10-prompt sampling-discipline probe (`min_p=0.05, top_p=0.8, top_k=20`) | 0 malformed | high malformed rate (repetition) | **0 malformed** |
| Long-prompt tok/s (k=1) | 34 | (was crashing before 0.20) | **105** |
| k=4 concurrent | stable | unstable on older builds | stable |

The deep fix recovers the spec-decoding speed-up without losing the sampling discipline that prevents repetition collapse on code models.

### testing

```
pytest tests/v1/sample/test_min_p_spec_decode.py -v
```

Three unit tests cover: no-op when no request has min_p>0, per-request masking with mixed num_draft_tokens, and that min_p=0 rows are untouched while a peer with min_p>0 still gets masked.

End-to-end validated against #42744's reproducer (Aider Polyglot Python harness with full anti-repetition stack + spec on).

Closes #42744

### CODE REVIEW

**Comment by camerono** [comment]
fyi context for the AGENTS.md re-evaluation:

this isn't a speculative fix. `RejectionSampler.apply_logits_processors` only iterates `logitsprocs.non_argmax_invariant`, which excludes `MinPLogitsProcessor` (`is_argmax_invariant() -> True`). result: min_p is silently dropped on every verified target token during spec decode — only the bonus token via the regular Sampler path honors it. that's a small but quiet correctness gap that's been masked by `_validate_spec_decode` 400'ing the requests instead of letting them through.

in-the-wild evidence: aider polyglot python (34 exercises) on Qwen3-Coder-Next NVFP4 + vllm 0.20 + this patch, spec off, full sampling stack (`min_p=0.05 / top_p=0.8 / top_k=20`) lands 38.2% pass@1 / 55.9% pass@2, 0 malformed responses. without the patch (under spec on) the model collapses into repetition loops on long completions because min_p never applies to the rejection-sampled tokens.

unit tests cover noop / per-row repeat_interleave / mixed zero-min_p peers. mirrors the existing `MinTokensLogitsProcessor.apply_with_spec_decode` pattern. dco signed.

ready for review when convenient.

**Comment by camerono** [comment]
It didn't have conflicts when I submitted this 60 days ago as courtesy support to the project . After I fix these, will I wait another 60 days and receive another merge conflict resolution request?

**Comment by camerono** [comment]
Rebased onto current main and green again. This is a real correctness fix — min_p was being silently dropped on verified target tokens during spec decode and only applied to the bonus token. On Qwen3-Coder-Next with spec decode enabled it takes throughput from 34 → 105 tok/s by eliminating the repetition collapse the missing min_p was causing. It's been open ~8 weeks and keeps drifting back into conflicts faster than it gets a review pass. @njhill @22quinn — could one of you take a look before it rots again?

**Comment by pocharlies** [comment]
Production validation from a second, independent deployment: this fix has been useful and correct for us.

Our setup is DeepSeek-V4-Flash-0731 on 2x NVIDIA DGX Spark GB10 (SM121a), vLLM `0.25.2.dev0+g752a3a504.d20260714`, TP=2, with DSpark speculative decoding configured as `k=5`, probabilistic verification.

Before applying this change, enabling `min_p=0.01` failed every `/v1/chat/completions` request at validation time with:

```
The min_p and logit_bias sampling parameters are not yet supported with speculative decoding.
```

We backported commit `1694a211e82ff047924bdfe807ad38b26da8da49` into our pinned runtime and selectively relaxed the request validator for `min_p` only; `logit_bias` remains rejected. We tested:

- per-request/per-position expansion for mixed speculative rows;
- the resulting mask against a direct min-p reference calculation;
- dispatch through `RejectionSampler.apply_logits_processors`;
- validator behavior (`min_p` accepted, `logit_bias` still blocked).

The three focused tests passed in the exact ARM64 production runtime. After deployment, a real request with explicit `min_p: 0.01` completed successfully:

```
HTTP 200
completion_tokens 1
finish_reason length
```

The server logs confirmed the request reached speculative decoding rather than bypassing it, with live DSpark acceptance metrics. One observed production window was:

```
Mean acceptance length: 3.61
Accepted: 47 tokens
Drafted: 90 tokens
Per-position acceptance rate: 0.833, 0.778, 0.444, 0.278, 0.278
Avg Draft acceptance rate: 52.2%
```

This is primarily correctness/support evidence, not a controlled performance claim: the request ran under concurrent production traffic and the first-token latency was high. The important result is that min-p was applied in the speculative path and the completion returned HTTP 200, whereas the unpatched runtime rejected the request before execution.

Thanks for submitting this. It closes a real gap for DSpark deployments that need min-p tail filtering.

AI assistance disclosure: AI tools assisted with investigation, test scaffolding, and drafting this report. I reviewed the patch behavior, ran the tests, and validated the deployment end-to-end.
